File: bogo2bogo/opencv/ImageViewer/ImageSpinner.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# ImageSpinner.py: cycle through images in a folder
import sys
import os
import pathlib
from PyQt5.QtCore import *
from PyQt5.QtGui import *
from PyQt5.QtWidgets import *


class ImageSpinner(QObject):
    def __init__(self, imagePath: str):
        assert (os.path.exists(imagePath)), f"ERROR: {imagePath} - invalid path!"
        self.currIndex = -1
        fileInfo = QFileInfo(imagePath)
        self.dir = fileInfo.absoluteDir()
        fileName = fileInfo.fileName()
        imageFilters = ["*.tiff", "*.tif", "*.jpg", "*.jpeg", "*.gif", "*.bmp", "*.png"]
        self.fileNames = self.dir.entryList(imageFilters, QDir.Files, QDir.Name)
        self.currIndex = self.fileNames.index(fileName)
        # self.currIndex = self.fileNames.index(
        #     QRegExp(QRegExp.escape(fileInfo.fileName()))
        # )

    def prevImagePath(self) -> str:
        self.currIndex = self.currIndex - 1
        if (self.currIndex < 0):
            self.currIndex = 0
        prevImagePath = self.dir.absolutePath() + QDir.separator() + \
            self.fileNames[self.currIndex]
        return prevImagePath

    def nextImagePath(self) -> str:
        self.currIndex = self.currIndex + 1
        if (self.currIndex > len(self.fileNames) - 1):
            self.currIndex = len(self.fileNames) - 1
        nextImagePath = self.dir.absolutePath() + QDir.separator() + \
            self.fileNames[self.currIndex]
        return nextImagePath

    def atFirstPath(self):
        return self.currIndex == 0

    def atLastPath(self):
        return self.currIndex == len(self.fileNames) - 1

File: bogo2bogo/ChocolafStyle/examples/C++/MyProjects/OpenCV/ImageViewer/ImageSpinner.cpp
#include "ImageSpinner.h"
#include <QtCore>
#include <QtGui>
#include <QtWidgets>
#include "common_funcs.h"
#include "ui_ImageViewer.h"

ImageSpinner::ImageSpinner(const QString &imagePath) : m_currIndex(-1)
{
  QFileInfo current(imagePath);
  m_dir = current.absoluteDir();
  QStringList imageFilters;
  imageFilters << "*.tiff"
               << "*.tif"
               << "*.jpg"
               << "*.jpeg"
               << "*.gif"
               << "*.bmp"
               << "*.png";
  m_fileNames = m_dir.entryList(imageFilters, QDir::Files, QDir::Name);
  m_currIndex = m_fileNames.indexOf(QRegExp(QRegExp::escape(current.fileName())));
}

QString ImageSpinner::prevImage()
{
  m_currIndex = m_currIndex - 1;
  if (m_currIndex < 0)
    m_currIndex = 0; // keep at 0th index
  QString prevImagePath =
      m_dir.absolutePath() + QDir::separator() + m_fileNames.at(m_currIndex);
  qDebug() << "Will display (prev): " << prevImagePath;
  // return dir.absoluteFilePath(m_fileNames.at(m_currIndex));
  return prevImagePath;
}

QString ImageSpinner::nextImage()
{
  m_currIndex = m_currIndex + 1;
  if (m_currIndex > m_fileNames.count() - 1)
    m_currIndex = m_fileNames.count() - 1; // keep at last index

  QString nextImagePath =
      m_dir.absolutePath() + QDir::separator() + m_fileNames.at(m_currIndex);
  qDebug() << "Will display (next): " << nextImagePath;
  // return dir.absoluteFilePath(m_fileNames.at(m_currIndex));
  return nextImagePath;
}

bool ImageSpinner::atFirst() const
{
    return m_currIndex == 0;
}

bool ImageSpinner::atLast() const
{
    return m_currIndex == m_fileNames.count() - 1;
}
